spinalcordtoolbox/utils/shell.py

```diff
@@ -98,6 +99,38 @@ def display_viewer_syntax(files, colormaps=[], minmax=[], opacities=[], mode='',
         printv(cmd + '\n', verbose=1, type='info')
 
 
+class SCTArgumentParser(argparse.ArgumentParser):
+    """Parser that centralizes initialization steps common across all SCT scripts."""
+    def __init__(self, *args, **kwargs):
+        def update_parent_default(key, value):
+            """A polite way of letting a child class have different default values than the parent class."""
+            # Source: https://stackoverflow.com/a/41623488
+            argspec = inspect.getfullargspec(super(SCTArgumentParser, self).__init__)
+            arg_index = argspec.args.index(key)
+            if len(args) < arg_index and key not in kwargs:
+                kwargs[key] = value
+
+        update_parent_default('formatter_class', SmartFormatter)
+
+        # Update "usage:" message to match how SCT scripts are actually called (no '.py')
+        frame = inspect.stack()[1]
+        module = inspect.getmodule(frame[0])
+        update_parent_default('prog', os.path.basename(module.__file__).strip(".py"))
+
+        # Disable "add_help", because it won't properly add '-h' to our custom argument groups
+        # (We use custom argument groups because of https://stackoverflow.com/a/24181138)
+        update_parent_default('add_help', False)
+
+        super(SCTArgumentParser, self).__init__(*args, **kwargs)
+
```

Comment:
It's inconsistency-prone to set `add_help=False` here but then expect every single script to add `--help` itself. I know the argparse API isn't great here -- there's an `add_` but no `get_` and even though there's titles to key on they're not in a dict. Upstream is

https://github.com/python/cpython/blob/056c08211b402b4dbc1530a9de9d00ad5309909f/Lib/argparse.py#L1436-L1439

```
    def add_argument_group(self, *args, **kwargs):
        group = _ArgumentGroup(self, *args, **kwargs)
        self._action_groups.append(group)
        return group
```

What if we turned the insert into an up-sert instead:

```suggestion

    def add_argument_group(self, title, *args, **kwargs):
         g = [g for g in parser._action_groups if g.title == title]
         if g:
             return g
         else:
             return super().add_argument_group(title, *args, **kwargs)
             
```

And added this to `__init__`:

```
    optional = self.add_argument_group("\nOPTIONAL ARGUMENTS")
    optional.add_argument(
        "-h",
        "--help",
        action="help",
        help="Show this help message and exit."
    )
```

then we could get rid of all the duplicate `--help`s, and plus the `--help` addition be on the same level as the `add_help=False` that makes room for it and nothing else has to change.

This is mildly buggy in that if a class *did* want to customize the --help group the cached version won't pick them up. I don't know how to handle that. Maybe add

```
if args or kwargs: raise ValueError
```
?

Reply:
Oh I just noticed you addressed this:

> In theory, we could also centralize the -h, -v, and -r arguments, too. But, that's a more involved effort, and can be done later as part of #2676.

I think that's a good idea too.

Reply:
Now that you mention it, `-h` is a lot easier to centralize compared to `-v` and `-r`. 

The reason I put it off was for the complexity of `-v` and `-r`, but I can definitely do `-h`. (Which also addresses https://github.com/neuropoly/spinalcordtoolbox/pull/3139#discussion_r549888651.)

Reply:
Thanks for pointing this out. :)

Reply:
> What if we turned the insert into an up-sert instead:
> 
> ```python
>     def add_argument_group(self, title, *args, **kwargs):
>          g = [g for g in parser._action_groups if g.title == title]
>          if g:
>              return g
>          else:
>              return super().add_argument_group(title, *args, **kwargs)             
> ```

The tricky part about this pattern is how sensitive it is to the specific `\nOPTIONAL ARGUMENTS` string. For example, I've seen some scripts use e.g. `\nMISC` instead, and in that case you would end up with 2 categories.

At the same time, this complexity is sort of inherent to our choice of using custom argument groups in the first place. :sweat_smile: 

TBH, since we're already messing with built-in argparse behavior, I wonder if we could just... _fix_ the whole "need for custom groups" quirk that `argparse` has.

Reply:
Hmm... these custom argument groups seem like they may a headache, and it seems tricky to quickly include the suggestion in this PR. I'm leaning toward what you said here:

> but it's okay for it to be in flux and a bit inconsistent temporarily, so long as it doesn't go long enough that we forget why it was like that. I'll link this thread over there so my investigation can help when we get to that.

I've added a TODO in 4cb11b6 to help not forget this. :sweat_smile:

Reply:
Great, je suis d'accord.